stl/inc/flat_set

```diff
@@ -208,11 +208,21 @@ public:
     // modifiers
     template <class... _Args>
     auto emplace(_Args&&... _Vals) {
-        return _Emplace(_Kty{_STD forward<_Args>(_Vals)...});
+        constexpr bool _Is_key_type = _In_place_key_extract_set<_Kty, remove_cvref_t<_Args>...>::_Extractable;
+        if constexpr (_Is_key_type) {
+            return _Emplace(_STD forward<_Args>(_Vals)...);

```

Comment:
As to `emplace` and `emplace_hint`:
When input is `(const)_Kty(&/&&)`, I believe there shouldn't be a temporary before insertion, even for `flat_multiset`; especially, for `flat_set` a `_Kty&&` input should not be moved to a temporary if insertion fails.
For example, the following code holds true for `set` and `unordered_set`
```cpp
    Set<shared_ptr<int>> s;
    auto a = shared_ptr<int>(new int);
    auto b = a, c = a;
    assert(a == b && a == c);
    assert(!(a < b) && !(b < a));
    s.insert(a);
    s.emplace(move(b));
    assert(b != nullptr);
    s.insert(move(c));
    assert(c != nullptr);
```